klippy/extras/load_cell_probe.py

```diff
@@ -0,0 +1,699 @@
+
+
+# ProbeSession that implements Tap logic
+class TapSession:
+    def __init__(self, config, tapping_move, probe_params_helper):
+        self._printer = config.get_printer()
+        self._tapping_move = tapping_move
+        self._probe_params_helper = probe_params_helper
+        self._activator = ProbeActivationHelper(config)
+        # Session state
+        self._results = []
+
+    def start_probe_session(self, gcmd):
+        self._activator.activate_probe()
+        return self
+
+    def end_probe_session(self):
+        self._activator.deactivate_probe()
+        self._results = []
+
+    # probe until a single good sample is returned or retries are exhausted
+    def run_probe(self, gcmd):
+        epos, is_good = self._tapping_move.run_tap(gcmd)
+        self._results.append(epos)
+
+    def pull_probed_results(self):
+        res = self._results
+        self._results = []
+        return res
+
+
+class LoadCellProbeCommands:
+    def __init__(self, config, load_cell_probing_move,
+                 tapping_move, config_helper):
+        self._printer = config.get_printer()
+        self._load_cell_probing_move = load_cell_probing_move
+        self._tapping_move = tapping_move
+        self._config_helper = config_helper
+        self._register_commands()
+
+    def _register_commands(self):
+        # Register commands
+        gcode = self._printer.lookup_object('gcode')
+        # for now there can only be 1 probe so no need for mux_command
+        gcode.register_command("LOAD_CELL_TAP",
+            self.cmd_LOAD_CELL_TAP, desc=self.cmd_LOAD_CELL_TAP_help)
+        gcode.register_command("LOAD_CELL_TEST_TAP",
+            self.cmd_LOAD_CELL_TEST_TAP, desc=self.cmd_LOAD_CELL_TEST_TAP_help)
+
+    cmd_LOAD_CELL_TAP_help="Perform a single tap"
+    def cmd_LOAD_CELL_TAP(self, gcmd):
+        self._tapping_move.run_tap(gcmd)
+
+    cmd_LOAD_CELL_TEST_TAP_help = "Tap the load cell probe to verify operation"
+    def cmd_LOAD_CELL_TEST_TAP(self, gcmd):
+        taps = gcmd.get_int("TAPS", 3, minval=1, maxval=10)
+        timeout = gcmd.get_float("TIMEOUT", 30., minval=1., maxval=120.)
+        gcmd.respond_info("Tap the load cell %s times:" % (taps,))
+        reactor = self._printer.get_reactor()
+        for i in range(0, taps):
+            result = self._load_cell_probing_move.probing_test(gcmd, timeout)
+            if result == 0.:
+                # notify of error, likely due to timeout
+                raise gcmd.error("Test timeout out")
+            gcmd.respond_info("Tap Detected!")
+            # give the user some time for their finger to move away
+            reactor.pause(reactor.monotonic() + 0.2)
+        gcmd.respond_info("Test complete, %s taps detected" % (taps,))
+
+
+class LoadCellPrinterProbe:
+    def __init__(self, config):
+        cfg_error = config.error
+        try:
+            import numpy as np
+        except:
+            raise cfg_error("[load_cell_probe] requires the NumPy module")
```

Comment:
It would be preferable not to introduce a dependency on numpy for trivial usages.  Unless I'm missing something, the three usages of numpy in the code can be replaced with:
* `tare_counts = sum([t[2] for t in tare_samples]) / float(len(tare_samples))`
* `self.time = [ns[0] for ns in nd_samples]`
* `self.force = [ns[1] for ns in nd_samples]`

I understand you'll want to use numpy as part of the more extensive tap analysis, and I'm fine with that.  It'll be easier to review the code if we add numpy as part of that code though.

I'll make this change as part of merging if you prefer.

Reply:
You can remove them if you want. I disagree with your premise. Those are optimizations that I put into the critical code path (where the user is waiting, toolhead is just sitting there) intentionally.

I've also made `np` a global, as that solves for the multiple imports comment. If you want to take them out, you might want to wait until after the next PR where the other 20 usages are introduced.